### PR TITLE
improvement on adding block

### LINKED ISSUE #308: Use block constructor when adding blocks
Delay initialization of block when adding and use constructor from registry instead of trying to look up the constructor by name.

### PR DESCRIPTION
**Pull request by DivadNojnarg**
Fix #308.

### CODE REVIEW

**Comment by DivadNojnarg** [comment]
@nbenn. I'd be tempted to move `add_block_ui` and the server part to the stack code. However, it must be done in a way that @JohnCoene can disable it since the app demo uses drag and drop and you don't need this logic.

File: inst/examples/add-blocks/app.R [hunk 1]
```diff
@@ -20,12 +20,11 @@ shiny::shinyApp(
       vals$new_block <- NULL
       # Always append to stack
       loc <- length(stack$blocks)
-      block <- available_blocks()[[input$selected_block]]
       # add_block expect the current stack, the block to add and its position
       # (NULL is fine for the position, in that case the block will
       # go at the end)
       vals$new_block <- list(
-        block = block,
+        block = input$selected_block,
```
Comment (DivadNojnarg) [inline inst/examples/add-blocks/app.R:27]: `input$selected_block` is a string returned by the radio input (`names(available_blocks())`). Then on the stack server side, we get the constructor with `available_blocks()[[new_block()$block]]`.

**Comment by JohnCoene** [comment]
I'm using this function in this PR https://github.com/blockr-org/blockr/pull/298

Can be please look at #298

**Comment by DivadNojnarg** [comment]
The code that was causing issue #293 (try adding a theme_block). This works whatever the block class we define in the constructors:

```r
devtools::load_all()
library(ggplot2)

new_ggplot_block <- function(data, ...){
  col_names <- function(data) colnames(data)
  x_default <- function(data) col_names(data)[1]
  y_default <- function(data) col_names(data)[2]
  fields <- list(
    x = new_select_field(x_default, col_names, type = "name"),
    y = new_select_field(y_default, col_names, type = "name")
  )
  expr <- quote(
    ggplot2::ggplot(mapping = aes(x = .(x), y = .(y)))
  )
  classes <- c("ggplot_block", "plot_block")
  new_block(fields = fields,
            expr = expr,
            class = classes, ...)
}

ggplot_block <- function(data, ...) {
  initialize_block(new_ggplot_block(data, ...), data)
}

new_geompoint_block <- function(data, ...) {
  new_block(
    fields = list(),
    expr = quote(
      geom_point()
    ),
    class = c("geompoint_block", "plot_layer_block", "plot_block"),
    ...
  )
}

geompoint_block <- function(data, ...) {
  initialize_block(new_geompoint_block(data, ...), data)
}

new_theme_block <- function(data, ...) {
  new_block(
    fields = list(
      theme = new_select_field(
        "theme_minimal",
        grep("^theme_.*$", ls("package:ggplot2"), perl = TRUE, value = TRUE),
        type = "name"
      )
    ),
    expr = quote(
      .(theme)()
    ),
    class = c("theme_block", "plot_layer_block", "plot_block"),
    ...
  )
}

theme_block <- function(data, ...) {
  initialize_block(new_theme_block(data, ...), data)
}

new_facet_block <- function(data, ...) {
  sel_cols <- function(data){
    colnames(data$data)
  }
  sel_cols_1 <- function(data){
    sel_cols(data)[1]
  }
  new_block(
    expr = quote(
      ggplot2::facet_wrap(facets = .(facet_var))
    ),
    fields = list(
      facet_var = new_select_field(sel_cols_1, sel_cols)
    ),
    class = c("facet_block", "plot_layer_block", "plot_block")
  )
}

facet_block <- function(data, ...) {
  initialize_block(new_facet_block(data, ...), data)
}

blockr::register_block(
  constructor = theme_block,
  name = "theme_block",
  description = "scatter with stats",
  classes = c("theme_block", "plot_block", "plot_layer_block"),
  input = "data.plot",
  output = "data.plot"
)

stack <- new_stack(
  data_block,
  ggplot_block,
  new_geompoint_block
)

library(shiny)

ui <- fluidPage(
  theme = bslib::bs_theme(version = 5),
  add_block_ui(),
  generate_ui(stack)
)

server <- function(input, output, session) {
  nb <- eventReactive(input$add, {
    list(
      block = input$selected_block,
      position = NULL
    )
  })
  generate_server(stack, new_block = nb)
}

shinyApp(ui, server)
```

**Comment by DivadNojnarg** [comment]
@JohnCoene: do you remember why we have 2 ways to add blocks in the stack server module:

```r
observeEvent(
        {
          req(new_block)
          new_block()
        },
        {
          add_block_stack(new_block()$block, new_block()$position, vals = vals)
        }
      )
```

and with the `add` button:

```r
observeEvent(input$add, {
        add_block_stack(
          block_to_add = available_blocks()[[input$selected_block]],
          position = NULL,
          vals = vals
        )
      })
```

Are we still using the first one?

File: inst/examples/stack/app.R [hunk 2]
```diff
@@ -0,0 +1,13 @@
+library(blockr)
```
Comment (DivadNojnarg) [inline inst/examples/stack/app.R:1]: Since the add block button is part of the stack, we update the example.

File: tests/testthat/test-stack.R [hunk 3]
```diff
@@ -39,7 +39,7 @@ test_that("stacks", {
 test_that("serve stacks", {
   skip_on_cran()
   driver <- shinytest2::AppDriver$new(
-    system.file("examples/add-blocks", package = "blockr"),
+    system.file("examples/stack", package = "blockr"),
```
Comment (DivadNojnarg) [inline tests/testthat/test-stack.R:42]: renamed to stack

File: R/stack.R [hunk 4]
```diff
@@ -182,10 +182,7 @@ add_block <- function(stack, block, position = NULL) {
   } else if (length(stack) > 1L) {
     data <- evaluate_block(stack[[1]])
     for (i in seq_along(stack)[-1L]) {
-      data <- evaluate_block(
-        do.call(class(stack[[i]])[[1]], list(data)),
-        data = data
-      )
+      data <- evaluate_block(stack[[i]], data = data)
```
Comment (DivadNojnarg) [inline R/stack.R:185]: @nbenn: main change of this PR -> we call the constructor get from `available_blocks()[[input$selected_block]]`.

File: inst/examples/add-blocks/app.R (hunk 1 above)
Comment (DivadNojnarg) [reply inst/examples/add-blocks/app.R]: Note: this isn't relevant anymore, this code was removed from the example as this happens within the stack server module.

**Comment by DivadNojnarg** [comment]
This one is good for checking @nbenn.

**Review by nbenn — APPROVED** [review]
LGTM, thanks!

**Pull request merged**